### DISCUSSION TITLE
Mutating an image using the Opacity extension has no effect

### DISCUSSION THREAD

**Discussion by jureles** [SixLabors/ImageSharp#2813] [Q&A]
I'm trying to apply transparency to an image by using the following code:

```
using var image = SixLabors.ImageSharp.Image.Load<Rgba32>(imageDataBytes);
image.Mutate(i => i.Opacity(0.2f));
image.SaveAsPng("D:\\Test.png");
```

Basically, what I'm trying to achieve is to modify the alpha channel of all image's pixels.

The problem is that the mutation has no effect for some images. `Image.PixelType.AlphaRepresentation` returns `null` even for images in which the Opacity mutation works as expected. If I don't specify pixel format and use the non-generic `Image.Load` then the loaded Image has a 24 bit pixel format - I'm just guessing if this might be a clue to the problem. This is the image that is causing the problem:
[sample_640×426.zip](https://github.com/user-attachments/files/17091357/sample_640x426.zip)

I need this to work for any (incoming) image of any supported image format. How do I properly make images semi-transparency - for any type of image?

I'm using ImageSharp v2.1.9.

Many thanks,
Jure

**Comment by tiesont** [comment]
Do you see the same behavior with the latest version? Version 2.1.9 is fairly old at this point, and only the v3 branch is receiving bugfixes (with some exceptions).

**Reply by jureles** [reply]
@tiesont thank you for the fast response. No, I haven't yet. I can try it with the latest 3.x version. But the problem is that our software runs inside MS Dynamics 365FO which means we are stuck with .NET Framework for years to come. This means we are stuck with ImageSharp v2.1.x.

I can try and run the latest release and let you know if I can reproduce the same behavior.

Still, I can confirm that this behavior is not consistent and only appears with some images (e.g., a BMP).

Cheers,
Jure

**Reply by tiesont** [reply]
Not an ImageSharp dev, to be clear, and I haven't had to work with BMP files at this point, so this is just a somewhat-educated guess based on my usage of ImageSharp.

If the incoming BMP is a 24-bit bitmap, then it doesn't have an alpha layer. That might be affecting how ImageSharp (especially an older version) is allocating resources. I'd have to dig into the code, though. 

Might be worth adding some examples of images which work and images which do not work - the easier it is for someone to reproduce what you're seeing, then the more likely it is for someone to come up with either a solution or an explanation of what's happening.

**Reply by jureles** [reply]
@tiesont, thank you for the input and idea. I updated my initial post.

**Reply by JimBobSquarePants** [reply]
Use an explicit `PngEncoder` instance when saving and set the `PngColorType` to `RgbWithAlpha`

**Reply by jureles** [reply]
@JimBobSquarePants Thank you so much for the solution. It works like a charm!

I know this is not the right place for giving feedback, but I will use this opportunity to tell you that we have almost completely (apart from WMF/EMF support) replaced System.Drawing with ImageSharp and it works like a charm. It is robust, and we haven't stumbled on any deficiency (compared to System.Drawing) apart from WMF/EMF formats, which I understand it doesn't make sense to cover with a modern library such as ImageSharp.

All in all, thank you so much for such a great library you are developing for the community - much appreciated! We have been supporting you for a while now, and we plan to increase support in the future as our team grows.

Best,
Jure

**Reply by JimBobSquarePants** [reply]
Feedback like this is always welcome! Thank you!
